feat: workspaces can manage tracked repositories now (search repos flow)

nickytonline opened the pull request:
## Description

This is under a feature flag atm. Only the OpenSauced team can see it. There is currently minimal workspace creation/edition in the app. This adds the ability to manage tracked repositories.

The search is finicky sometime when searching for a repository. I thought this was something I introduced, but appears to be happening elsewhere in the app as well, e.g. adding repositories in the new insights page, so that will get addressed in a separate issue.

Note that this PR is only for the search by repos flow. Another pull request will introduce search for repost by org (#2505).

![image](https://github.com/open-sauced/app/assets/833231/50c5d85c-0ef3-4a1d-b955-2d65656920a2)

There will probably be some additional polish to the whole epic for tracked repositories, but that will be small PRs coming after.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Relates #2437
Closes #2482


## Mobile & Desktop Screenshots/Recordings

![CleanShot 2024-01-23 at 23 32 14](https://github.com/open-sauced/app/assets/833231/16609611-8dd2-4e07-afc3-8016444ef6c5)

![CleanShot 2024-01-23 at 23 43 48](https://github.com/open-sauced/app/assets/833231/139a5068-f811-45b2-810f-f01e2539998a)

![CleanShot 2024-01-23 at 23 47 30](https://github.com/open-sauced/app/assets/833231/3c8db3bb-1d5e-4e01-bfa4-c55698e8d936)

![CleanShot 2024-01-23 at 23 46 56](https://github.com/open-sauced/app/assets/833231/6da5a66b-0c08-41b4-960d-bc4760415bc9)

https://github.com/open-sauced/app/assets/833231/786f3872-7691-4ccd-9453-0dae52674bf0

## TODOs

- [x] polishing of the feature
- [x] get useSWR caching issue sorted for tracked repos on the settings page. The first call should have repos in the settings page after creating a new workspace
- [x] width resizing of wizard when searching for repos and results load in the table
- [x] ~Get search component in search by repos step working properly. Results don't always load~ This appears to be an existing issue in the app.
- [x] Get search component in search by repos step working properly for clicking on a selected repo. Currently clicking closes the search, but does not add the selected repo. Probably missing an onClick
- [x] Fix Storybook stories and most likely merge them into one story for wizard steps.


For these, they're a separate issue raised here https://github.com/open-sauced/api/issues/511

- [ ] ~sometimes creating a repo fails because the endpoint for adding repos to the new workspace gives a 404. The workspace is created successfully, so it does exist. My guess is the endpoint 404s because maybe the database hasn't updated yet, so the repos endpoint 404s. A second attempt could be made as a "temporary" fix~
- [ ] ~sometimes I get 404s for the updating repos endpoint. Same for getting repos, e.g. 
https://beta.api.opensauced.pizza/v1/workspaces/9e0c76a8-9034-4670-94e8-f73b4a5c8679/repos~
- [ ] ~separate issue, but make the sidebar collapsible~

## Steps to QA

1. Ensure you're logged in
2. Navigate to https://deploy-preview-2502--oss-insights.netlify.app/workspaces/new
3. Fill in the required workspace info (name) and add some repositories to track.
4. Create the workspace and you'll be redirected to the workspace settings page.
5. Add or remove some tracked repos and update the workspace.
6. Notice the tracked repositories update. They might rearrange in order.


## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?

We'll need to coordinate with @BekahHW for documentation updates around workspaces, list and insights.

## [optional] What gif best describes this PR or how it makes you feel?

<img src="https://media4.giphy.com/media/3ohhweiVB36rAlqVCE/giphy.gif"/>

<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

nickytonline commented on lib/utils/github-api-fetcher.ts:
```diff
@@ -0,0 +1,31 @@
+import { Fetcher } from "swr";
```
A fetcher for SWR so that we get cached results for repo searches.

nickytonline commented on pages/workspaces/[workspaceId]/settings.tsx:
```diff
@@ -20,23 +22,58 @@ async function saveWorkspace({
+    pathValidator: () => true,
+  });
+
+  const [{ data, error }, { data: repoData, error: reposError }] = await Promise.all([
```
Since we already have the workspace ID, we can updated these at the same time.

nickytonline commented on pages/workspaces/[workspaceId]/settings.tsx:
```diff
@@ -79,87 +116,149 @@ const WorkspaceSettings = ({ workspace }: WorkspaceSettingsProps) => {
+          onRemoveTrackedRepo={(event) => {
+            const { repo } = event.currentTarget.dataset;
+
+            if (!repo) {
```
We should never end up in this `if` block. This could be removed and switch

```typescript
const { repo } = event.currentTarget.dataset;
```

to

```typescript
const repo = event.currentTarget.dataset.repo!;
```

nickytonline commented on pages/workspaces/new.tsx:
```diff
@@ -1,95 +1,153 @@
+  repos: { full_name: string }[];
 }) {
-  const { data, error } = await fetchApiData<Workspace>({
+  // TODO: check if the workspace ID already exists in case the repos part doesn't work
```
TODO: Nick

nickytonline commented:
I know we're meeting up tomorrow @bdougie, but if you want to take this for a test run...

There's a couple bumps in it still. Mainly the search results not always populating (what I'm looking into atm). And the search by org for tracked repos will be in a separate PR.

bdougie commented:
> I know we're meeting up tomorrow @bdougie, but if you want to take this for a test run...
> 
> There's a couple bumps in it still. Mainly the search results not always populating (what I'm looking into atm). And the search by org for tracked repos will be in a separate PR. Also, the GH API isn't always fast when searching for repos. I use `useSWR` for the queries, so previous queries tend to pop up quicker as they're cached.

Tried it our and successful tracked some repos. The one thing I noticed is I didn't get `open-sauced/app` to show when looking for `open-sauced`. I imagine there is some finessing you are still doing there as well. 

<img width="828" alt="Screen Shot 2024-01-22 at 4 38 22 PM" src="https://github.com/open-sauced/app/assets/5713670/f256750c-ee7e-4564-8768-d957b80c30ac">

nickytonline commented on lib/utils/workspace-utils.ts:
```diff
@@ -0,0 +1,109 @@
+  });
+
+  if (workspace) {
+    const { data: repoData, error: reposError } = await fetchApiData<any[]>({
```
@jpmcb, could the `/workspaces` creation endpoint take repos as well? Just asking as I've run into this a few times where the workspace gets created, but then adding repositories fail. Since it's not all in a transaction, the workspace exists, but a retry creates a new workspace when the repos get added successfully in this call, `workspaces/${workspace.id}/repos`.

jpmcb replied:
Sure thing: we can add repos directly to workspace creation.

> Since it's not all in a transaction, the workspace exists, but a retry creates a new workspace when the repos get added successfully in this call, `workspaces/${workspace.id}/repos`

I'm confused, why would the retry here create a new workspace if creating a new workspace and adding repos to it are separate?

jpmcb replied:
Hmmm I'm second guessing adding a repos payload to workspace creation: what if there are invalid repos in the payload for creating a workspace? Would the whole call fail? Should the workspace still be created but no repos be added? Only the valid repos be added with an error bubbled up for the repos that are invalid?

In my mind, two separate calls (one for creating a workspace and another one for mutating the repos within a workspace) makes logical sense to me.

nickytonline replied:
This is the scenario. Say the workspace gets created (first call) and the second call to add repos fails. The workspace has been created, but we're still on the creation page. The user currently sees an error if the workspace (first call) fails or if the repos fail to add (second call).

So say they retry, the create workspace (first call) runs again (successfully), and then the second call finally works, they're redirected to the settings page but now they have an additional workspace (the first one that created the workspace but failed to create the repos in the second call).

I can guard against this by storing the workspaceId  client-side if the first part was successful, but then that seems kind of hacky. This is why I was proposing allowing adding repos on creation. I'd rather it all fail then have one part pass.

The weird thing is what I mentioned to you the other day. Sometimes when the creation of repos fail, they're valid repos (since they come from the search repos endpoint from the GitHub API. Or do we 404 on invalid repos sometimes because a valid repo is a repo that we've indexed with pizza time?

nickytonline replied:
It relates to my TODOS above in the PR description.

> - [ ] sometimes creating a repo fails because the endpoint for adding repos to the new workspace gives a 404. The workspace is created successfully, so it does exist. My guess is the endpoint 404s because maybe the database hasn't updated yet, so the repos endpoint 404s. A second attempt could be made as a "temporary" fix
> - [ ] sometimes I get 404s for the updating repos endpoint. Same for getting repos, e.g. 
https://beta.api.opensauced.pizza/v1/workspaces/9e0c76a8-9034-4670-94e8-f73b4a5c8679/repos

jpmcb replied:
Sure thing: either way, let's move forward with allowing clients to have a `repos` payload: https://github.com/open-sauced/api/issues/509

I'll stub out some work and get that in today!

---

> Sometimes when the creation of repos fail, they're valid repos

Hmm very weird, still not sure why this would occur: if we can reliably reproduce it, let's add those steps to an issue

nickytonline replied:
Let's keep them as is for now as you mentioned it's something we need to sort out in https://github.com/open-sauced/api/issues/511

nickytonline commented on components/Workspaces/TrackedRepoWizard/PickReposOrOrgStep.tsx:
```diff
@@ -1,30 +1,25 @@
+export const PickReposOrOrgStep = ({ onSearchRepos, onImportOrg }: PickReposOrOrgStepProps) => {
+  const buttonRef = useRef<HTMLButtonElement>(null);
+
+  useEffectOnce(() => {
```
this gives focus to the search by repos button.

nickytonline commented on components/Workspaces/TrackedRepoWizard/SearchByReposStep.tsx:
```diff
@@ -34,52 +28,79 @@ const EmptyState = () => {
+  const [searchIsLoading, setSearchIsLoading] = useState(false);
+
+  useEffectOnce(() => {
+    (formRef.current?.querySelector('[name="query"]') as HTMLInputElement)?.focus();
```
gives focus to the search bar when you're on the search repos step of the tracked repo wizard.

nickytonline commented on components/Workspaces/TrackedRepoWizard/TrackedRepoWizard.tsx:
```diff
@@ -0,0 +1,91 @@
+            suggestedRepos={suggestedRepos}
+          />
+        );
+      // TODO: other steps
```
The other step will be coming up in #2505.

nickytonline commented on lib/hooks/api/useGetWorkspaceRepositories.ts:
```diff
@@ -0,0 +1,29 @@
+import useSWR, { Fetcher } from "swr";
```
Leveragin useSWR so that tracked repositories can be cached.

nickytonline commented on lib/hooks/useSearchRepos.ts:
```diff
@@ -0,0 +1,27 @@
+import useSWR, { Fetcher } from "swr";
```
Leveragin useSWR so that repository queries can be cached. It would be good to leverage this for repository search elsewhere in the app, e.g. insights pages.

bdougie commented:
@nickytonline and @isabensusan when reviewing I added repos and then left the page. I realized eventually that I never hit "update workspace." My suggestion would be add an "are you sure" pop up if someone leaves without clicking update.  

If I had added more than 2 repos one by one, it would have been annoying to do that again.

![Screen Shot 2024-01-23 at 3 27 30 PM](https://github.com/open-sauced/app/assets/5713670/c3082be2-f7f2-4e8f-87f8-7e4888d6f28a)

NsdHSO commented on pages/workspaces/new.tsx:
```diff
@@ -30,7 +30,7 @@ const NewWorkspace = () => {
       members: [],
       sessionToken: sessionToken!,
-      repos: trackedRepos.map((repo) => ({ full_name: repo })),
+      repos: Array.from(trackedRepos).map((repo) => ({ full_name: repo })),
```
`Array.from` can take a mapping function as its second argument.

nickytonline replied:
Good call. I forgot about that. Thanks @NsdHSO!

nickytonline commented on components/Workspaces/SearchReposTable.tsx:
```diff
@@ -75,16 +58,22 @@ export const SearchedReposTable = ({
+                        onChange={(event) => {
+                          onSelect("");
+                        }}
+                        // TODO: handle checked/unchecked
```
Looking at this now. 😅

brandonroberts commented:
The workflow worked well for me. I think we should debounce the repository search box, and potentially hit our API first, then GitHubs. 

It seems the GitHub token provided only gives you 10 requests before you hit the quota. If you logout/login you get another batch.

jpmcb reviewed (approved):
Looks amazing!!

---

I get abunch of github 403 when I try to search for repos and then it just spins forever:

<img width="756" alt="Screenshot 2024-01-24 at 12 29 49 PM" src="https://github.com/open-sauced/app/assets/23109390/eb740ec2-9046-4046-9f1c-4de75fdf9a2a">

Here's the message:

```
message	"API rate limit exceeded for 50.250.158.186. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)"
```

Which is interesting since I believe `50.250.158.186` is the forward IP to my VPN I'm using. I'm not seeing an authenticated token in the request, like Brandon said, a big enhancement would be using the users authenticated token!

nickytonline commented:
> @nickytonline and @isabensusan when reviewing I added repos and then left the page. I realized eventually that I never hit "update workspace." My suggestion would be add an "are you sure" pop up if someone leaves without clicking update.
> 
> If I had added more than 2 repos one by one, it would have been annoying to do that again.
> 
> ![Screen Shot 2024-01-23 at 3 27 30 PM](https://private-user-images.githubusercontent.com/5713670/299130495-c3082be2-f7f2-4e8f-87f8-7e4888d6f28a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYxMjU4MTQsIm5iZiI6MTcwNjEyNTUxNCwicGF0aCI6Ii81NzEzNjcwLzI5OTEzMDQ5NS1jMzA4MmJlMi1mN2YyLTRlOGYtODdmOC03ZTQ4ODhkNmYyOGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjRUMTk0NTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmEwOTQyNzRjZDg3ZjMzZWE5Y2M4ZmI3NzU0OWNmYzlmNDI2MmViNjUzNzRhOGJiNjZjZGNmYjA2OTgzMWQ4NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tJX7y9gTq3_Tzh5cLvBq_jWV8-IX_Gllp3Fc3RZLF0w)

I'm going to have to dig into this a bit more in a follow up PR. I used to use `onbeforeunload` on the window for stuff like this, but it doesn't appear to behave how it used to. It only stops you from leaving the page if you try to refresh the page.

bdougie reviewed (approved):
Fine to see this merge and do follow ups PRs for the things brought

nickytonline commented:
#2483 might fix the odd behaviour of search I was experiencing while working on this PR.
